lib/LANraragi/Plugin/Metadata/MEMS.pm

```diff
@@ -0,0 +1,161 @@
+  if ($lrr_info->{oneshot_param} =~ /e(?:x|-)hentai\.org\/g\/(\d+)\/([0-9a-z]+)/i) {
+    $gallery_id = $1;
+    $gallery_token = $2;
+    $logger->debug("Directly using gallery ID $gallery_id and token $gallery_token from oneshot parameters.");
+  } elsif ($lrr_info->{existing_tags} =~ /source:e(?:x|-)hentai\.org\/g\/(\d+)\/([0-9a-z]+)/i) {
+    $gallery_id = $1;
+    $gallery_token = $2;
+    $logger->debug("Directly using gallery ID $gallery_id and token $gallery_token from source tag.");
+  } else {
+    # Use the gallery ID and token in the filename to directly locate the gallery. Note that the regex does not have "$"
+    # at the end, so the filename can have other information attached after the identifiers.
+    ($gallery_id, $gallery_token) = ($lrr_info->{archive_title} =~ /.+? \[GID (\d+) GT ([0-9a-z]+)\]/);
+    if ($gallery_id eq '' || $gallery_token eq '') {
+      my $file_error = "Skipping archive without connecting to EH, because the archive title does not have valid "
+        . "gallery identifiers from Mayriad\'s EH Master Script.";
+      $logger->error($file_error);
+      return (error => $file_error);
+    }
+  }
+
+  # Retrieve metadata directly using EH API.
+  my ($eh_all_tags, $eh_title, $error) = &get_eh_metadata($gallery_id, $gallery_token, $save_jpn_title,
+    $save_additional_metadata);
+  if ($error ne '') {
+    $logger->error($error);
+    return (error => $error);
+  }
+
+  # Put all metadata together and add them to LRR.
+  $logger->info("Sending the following tags to LRR: $eh_all_tags");
+  my %metadata = (tags => $eh_all_tags);
+  # Add the source tag outside get_eh_metadata(), so that this tag is only added when metadata has been successfully
+  # retrieved; otherwise $metadata{tags} may only contain this source tag and truly untagged galleries would be
+  # incorrectly hidden.
+  my $host = ($use_exhentai ? 'exhentai.org' : 'e-hentai.org');
+  $metadata{tags} .= ", source:$host/g/$gallery_id/$gallery_token";
+  # Title is always updated to hide the identifiers and also to reflect title changes due to rename petitions.
+  $metadata{title} = $eh_title;
+  # Return a hash containing the new metadata to be added to LRR.
+  return %metadata;
+}
+
+# get_eh_metadata($gallery_id, $gallery_token, $save_jpn_title, $save_additional_metadata)
+# Sends an EH API gallery metadata request and extracts metadata from the returned JSON.
+sub get_eh_metadata {
+  my ($gallery_id, $gallery_token, $save_jpn_title, $save_additional_metadata) = @_;
+  my $api = 'https://api.e-hentai.org/api.php';
+  my $user_agent = Mojo::UserAgent->new;
+  my $logger = get_logger('Mayriad\'s EH Master Script', 'plugins');
+
+  # Send the API request.
+  my $response = $user_agent->post(
+    $api => json => {
+      method => 'gdata',
+      gidlist => [[$gallery_id, $gallery_token]],
+      namespace => 1
+    }
+  )->result;
+
+  # Check for temporary ban and warning messages.

```

Comment:
There's no rate limits on the API that I know of unlike the rest of the site, so this check/warning is probably unnecessary here.

Reply:
The EH wiki says the API has a load limit ("25 entries per request, 4-5 sequential requests usually okay before having to wait for ~5 seconds") and an EH mod mentioned API related bans to me once, and the stock plugin also has these warnings, so I added them here.

However, I did not get banned while testing this plugin with 0 delay, so maybe the way LANraragi sends requests one at a time is safe from bans, or maybe there is no real limit? I do not mind removing these, and we can remove them from the stock plugin as well.

Reply:
Didn't think we had those in place for the API calls as well in the stock plugin -- I guess they can stay but if you're going for option 2 this point is moot.

Reply:
It seems the API does have a load limit. See [this forum post](https://forums.e-hentai.org/index.php?s=&showtopic=238939&view=findpost&p=5694807).